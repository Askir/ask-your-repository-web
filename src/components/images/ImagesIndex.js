// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import InfiniteScroll from 'react-infinite-scroller';
import Gallery from './Gallery';
import ImageService from '../../services/ImageService';
import type { Team } from '../../models/Team';
import type { Image } from '../../models/Image';
import type { AppState } from '../../state/AppState';
import './ImagesIndex.scss';

type Props = {
  activeTeam: ?Team,
};

type State = {
  images: Array<Image>,
  offset: number,
  endReached: boolean,
};

export const limit = 12;

class ImagesIndex extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      images: [],
      offset: 0,
      endReached: false,
    };
  }

  componentDidMount() {
    this.loadMoreImages();
  }

  loadMoreImages = async () => {
    if (this.state.endReached) return;

    const currentOffset = this.state.offset;
    this.increaseOffset();

    if (!this.props.activeTeam) return;

    try {
      const params = {
        teamId: this.props.activeTeam.id,
        offset: currentOffset,
        limit,
      };
      const images = await ImageService.list(params);
      this.receiveImages(images);
    } catch (error) {
      // TODO: Handle error
    }
  }

  increaseOffset() {
    const { offset } = this.state;
    this.setState({ offset: offset + limit });
  }

  receiveImages(fetchedImages: Image[]) {
    const { images } = this.state;

    this.setState({ images: [...images, ...fetchedImages] });

    if (fetchedImages.length === 0) {
      this.setState({ endReached: true });
    }
  }

  render() {
    const { images } = this.state;

    return (
      <div className="ImagesIndex">
        <InfiniteScroll
          initialLoad={false}
          hasMore={!this.state.endReached}
          loadMore={this.loadMoreImages}
        >
          <Gallery images={images} />
        </InfiniteScroll>
      </div>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  activeTeam: state.activeTeam,
});

export default connect(mapStateToProps)(ImagesIndex);
