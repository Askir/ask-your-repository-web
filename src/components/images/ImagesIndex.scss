.ImagesIndex {
	max-width: var(--max-width);
	height: 100%;
	margin: 30px auto;
	display: flex;
	align-items: center;
	justify-content: center;
}
