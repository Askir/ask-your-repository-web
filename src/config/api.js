import axios from 'axios';
import qs from 'qs';

const api = axios.create({
  baseURL: process.env.REACT_APP_API_URL,
  requestInterceptor: config => ({ ...config }),
  paramsSerializer: parameters => qs.stringify(parameters, { arrayFormat: 'repeat' }),
});

export default api;
