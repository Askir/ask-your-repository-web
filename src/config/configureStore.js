// @flow
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../state/rootReducer';

function configureStore() {
  let middleware = [thunk];

  if (process.env.NODE_ENV !== 'production') {
    const { logger } = require('redux-logger');
    middleware = [...middleware, logger];
  }

  return createStore(
    rootReducer,
    applyMiddleware(...middleware),
  );
}

export default configureStore;
