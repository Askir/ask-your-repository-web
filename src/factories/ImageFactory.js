// @flow
import _ from 'lodash';
import faker from 'faker';
import TagFactory from './TagFactory';
import type { Image } from '../models/Image';

class ImageFactory {
  static createDummyImage(tagAmount: number = 3): Image {
    return {
      id: faker.random.number(),
      src: faker.image.dataUri(),
      tags: _.times(tagAmount, () => TagFactory.createDummyTag()),
    };
  }
}

export default ImageFactory;
