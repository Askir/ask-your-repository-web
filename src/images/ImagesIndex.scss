.IndexView {
	width: 100%;
	height: 100%;
	position: relative;
	margin-top: 100px;
}

.ImagesPage {
	max-width: var(--max-width);
	height: 100%;
	margin: 0 auto;
	display: flex;
	align-items: center;
	justify-content: center;
}

@media (min-width: 1200px) {

  :root {
    --max-width: 1200px;
  }

}

@media (min-width: 1025px) and (max-width: 1280px) {

  :root {
    --max-width: 90vw;
  }

}

@media (min-width: 768px) and (max-width: 1024px) {

  :root {
    --max-width: 90vw;
  }

}

@media (min-width: 330px) and (max-width: 767px) {

  :root {
    --max-width: 80vw;
  }

}
