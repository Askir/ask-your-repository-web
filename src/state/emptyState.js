// @flow
import type { AppState } from './AppState';

const emptyState: AppState = {
  image: null,
  images: [],
};

export default emptyState;
