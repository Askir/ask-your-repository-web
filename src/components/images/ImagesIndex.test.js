// @flow
import React from 'react';
import thunk from 'redux-thunk';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import ImagesIndex from './ImagesIndex';
import api from '../../config/api';
import emptyState from '../../state/emptyState';
import ImageFactory from '../../factories/ImageFactory';

const mockStore = configureStore([thunk]);

jest.mock('../../config/api');

describe('<ImagesIndex />', () => {
  let store;
  let wrapper;

  beforeAll(() => {
    api.get.mockImplementation(() => Promise.resolve({ data: { images: [] } }));
  });

  beforeEach(() => {
    store = mockStore(emptyState);
    wrapper = shallow(<ImagesIndex store={store} />).dive();
  });

  afterEach(() => {
    api.get.mockClear();
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('server responding with a non empty array', () => {
    beforeAll(() => {
      const images = [ImageFactory.createDummyImage()];
      api.get.mockImplementation(() => Promise.resolve({ data: { images } }));
    });

    it('should increase the offset by the limit', async () => {
      expect(wrapper.state().offset).toEqual(10);
    });
  });

  it('should not send a request when end is reached', async () => {
    const calls = api.get.mock.calls.length;
    wrapper.setState({ endReached: true });
    await wrapper.instance().loadMoreImages();
    expect(api.get).toHaveBeenCalledTimes(calls);
  });

  describe('server responding with empty array', () => {
    beforeAll(() => {
      api.get.mockImplementation(() => Promise.resolve({ data: { images: [] } }));
    });

    it('should set state to end reached', async () => {
      expect(wrapper.state().endReached).toEqual(true);
    });
  });
});
