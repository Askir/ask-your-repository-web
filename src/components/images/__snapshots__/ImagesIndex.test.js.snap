// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImagesIndex /> renders correctly 1`] = `
<ImagesIndex
  activeTeam={
    Object {
      "id": "2dedf601-f0bf-4290-ac3c-f5761234d068",
      "members": Array [
        Object {
          "email": "test@example.com",
          "id": "f1bc8f40-7698-4df7-b410-efac1f1a03d2",
          "username": "test_user",
        },
      ],
      "name": "odio expedita aut",
    }
  }
  dispatch={[Function]}
  location={
    Object {
      "search": " ",
    }
  }
>
  <div
    className="ImagesIndex"
  >
    <InfiniteScroll
      element="div"
      getScrollParent={null}
      hasMore={true}
      initialLoad={false}
      isReverse={false}
      loadMore={[Function]}
      loader={null}
      pageStart={0}
      ref={null}
      threshold={250}
      useCapture={false}
      useWindow={true}
    >
      <div>
        <Gallery
          images={Array []}
        >
          <div
            className="Gallery"
          />
        </Gallery>
      </div>
    </InfiniteScroll>
  </div>
</ImagesIndex>
`;
